File: 2019/Python/day11.py
### =============================================
### [ ADVENT OF CODE ] (https://adventofcode.com)
### 2019 - Mina Pêcheux: Python version
### ---------------------------------------------
### Day 11: Space Police
### =============================================
from intcode import IntcodeProgram

# [ Input parsing functions ]
# ---------------------------
def parse_input(data):
    '''Parses the incoming data into processable inputs.
    
    :param data: Provided problem data.
    :type data: str
    :return: Parsed data.
    :rtype: list(int)
    '''
    return [ int(x) for x in data.split(',') if x != '' ]

# [ Computation functions ]
# -------------------------
### Part I + II
def process_inputs(inputs, start_white=False, display=False, debug=False):
    '''Executes the Intcode program on the provided inputs and finds out the
    number of panels that have been painted at least once. It can also display
    the message that has been painted if necessary.
    
    :param inputs: List of integers to execute as an Intcode program.
    :type inputs: list(int)
    :param start_white: If true, then the starting panel is considered painted
        white. Else, it is considered painted black.
    :type start_white: bool
    :param display: If true, then the final state of the board is displayed in
        the shell.
    :type display: bool
    :param debug: Whether or not the IntcodeProgram should debug its
        execution at each instruction processing.
    :type debug: bool
    :return: Number of panels painted at least once.
    :rtype: int
    '''
    # prepare the board and the set of painted panels:
    # - the board only remembers the panels painted white
    # - the painted set remembers all the panels that have been painted at least
    # once
    board = set()
    painted = set()
    # initialize the painting robot: facing up, at the origin coordinates
    dir = 0
    x, y = 0, 0
    # (if starting white: mark the current panel as already painted white)
    if start_white:
        board.add((x, y))

    # prepare the program instance to read the given inputs as an Intcode
    # program
    program = IntcodeProgram(inputs, debug=debug)
    running = True
    # execute the program until it halts (but pause every 2 outputs)
    while running:
        # get the input depending on the state of the panel: if painted white
        # (i.e. visible in the board), the input is 1; else it is 0
        input = 1 if (x, y) in board else 0
        # insert the input in the program's memory
        program.push_memory(input)
        # execute until 2 digits have been outputted
        state = program.run(pause_every=2)
        # check for state:
        # . if paused: parse outputs and apply the actions
        if state == 'pause':
            color, rotation = program.output
            program.reset_output()
            if color == 1:
                board.add((x, y))
            else:
                board.discard((x, y))
            painted.add((x, y))
            m = -1 if rotation == 0 else 1
            dir = (dir + m) % 4
            if dir == 0: # up
                y -= 1
            elif dir == 1: # right
                x += 1
            elif dir == 2: # down
                y += 1
            elif dir == 3: # left
                x -= 1
        # . else: stop the program
        elif state is None:
            running = False
            break
            
    # if necessary, display the final message, i.e. the board that
    # has been printed (and only contains the painted panels)
    if display:
        # . separate horizontal from vertical coordinates
        x, y = zip(*board)
        # . find board boundaries
        min_x, max_x = min(x), max(x)
        min_y, max_y = min(y), max(y)
        # . iterate through the board to print out the message
        print('')
        for y in range(min_y, max_y + 1):
            row = ''
            for x in range(min_x, max_x + 1):
                if (x, y) in board:
                    row += '█'
                else:
                    row += ' '
            print(row)
        print('')
        
    # get the number of panels that have been painted at least once
    # (counts each panel once, and counts the panels even if they have been
    # repainted black)
    return len(painted)

if __name__ == '__main__':
    # get input data
    data_path = '../data/day11.txt'
    inputs = parse_input(open(data_path, 'r').read())
    
    ### PART I
    solution = process_inputs(inputs)
    print('PART I: solution = {}'.format(solution))
    
    ### PART II
    solution = process_inputs(inputs, start_white=True, display=True)
    print('PART II: solution (see above in the shell)')
    

File: 2019/Javascript/day11.js
/*
 * ================================================
 * [ ADVENT OF CODE ] (https://adventofcode.com)
 * 2019 - Mina Pêcheux: Javascript (NodeJS) version
 * ------------------------------------------------
 * Day 11: Space Police
 * ================================================
 */
'use strict'

const _ = require('lodash')
const fs = require('fs')
const expect = require('chai').expect
const { IntcodeProgram } = require('./utils/intcode')

// [ Input parsing functions ]
// ---------------------------
/**
 * Parses the incoming data into processable inputs.
 * @param {string} data - Provided problem data.
 * @returns {array(int)} - Parsed data.
 */
const parseInput = (data) => {
  return _.map(_.filter(_.split(data, ','), (l) => l.length > 0), Number)
}

// [ Computation functions ]
// -------------------------
/* PART I + II */
/**
 * Executes the Intcode program on the provided inputs and finds out the
   number of panels that have been painted at least once. It can also display
   the message that has been painted if necessary.
 * @param {array(int)} inputs - List of integers to execute as an Intcode program.
 * @param {bool} startWhite - If true, then the starting panel is considered
 *                            painted white. Else, it is considered painted black.
 * @param {bool} display - If true, then the final state of the board is displayed
 *                         in the shell.
 * @param {bool} debug - Whether or not the IntcodeProgram should debug its
 *                       execution at each instruction processing.
 * @returns {int} - Number of panels painted at least once.
 */
const processInputs = (inputs, startWhite=false, display=false, debug=false) => {
  // prepare the board and the set of painted panels:
  // - the board only remembers the panels painted white
  // - the painted set remembers all the panels that have been painted at least
  // once
  const board = new Set([])
  const painted = new Set([])
  // initialize the painting robot: facing up, at the origin coordinates
  let dir = 0
  let x = 0, y = 0
  // (if starting white: mark the current panel as already painted white)
  if (startWhite) {
    board.add(`${x},${y}`)
  }
  
  // prepare the program instance to read the given inputs as an Intcode program
  const program = new IntcodeProgram(inputs, debug)
  let running = true
  let input, state, color, rotation, m
  // execute the program until it halts (but pause every 2 outputs)
  while (running) {
    // get the input depending on the state of the panel: if painted white
    // (i.e. visible in the board), the input is 1; else it is 0
    input = (board.has(`${x},${y}`)) ? 1 : 0
    // insert the input in the program's memory
    program.pushMemory(input)
    // execute until 2 digits have been outputted
    state = program.run(2)
    // check for state:
    // . if paused: parse outputs and apply the actions
    if (state === 'pause') {
      color = program.output[program.output.length - 2]
      rotation = program.output[program.output.length - 1]
      program.resetOutput()
      if (color === 1) {
        board.add(`${x},${y}`)
      } else {
        board.delete(`${x},${y}`)
      }
      painted.add(`${x},${y}`)
      m = (rotation === 0) ? -1 : 1
      dir += m
      dir = ((dir % 4) + 4) % 4
      switch (dir) {
        case 0: // up
          y--
          break
        case 1: // right
          x++
          break
        case 2: // down
          y++
          break
        case 3: // left
          x--
          break
        default:
          break
      }
    // . else: stop the program
    } else if (state === null) {
      running = false
      break
    }
  }
  
  // if necessary, display the final message, i.e. the board that
  // has been printed (and only contains the painted panels)
  if (display) {
    const marker = '█'
    // . separate horizontal from vertical coordinates
    x = [], y = []
    board.forEach((coord) => {
      const [ xx, yy ] = _.split(coord, ',')
      x.push(parseInt(xx))
      y.push(parseInt(yy))
    })
    // . find board boundaries
    const minX = _.min(x), maxX = _.max(x)
    const minY = _.min(y), maxY = _.max(y)
    // . iterate through the board to print out the message
    console.log('')
    let row
    _.each(_.range(minY, maxY + 1), (y) => {
      row = ''
      _.each(_.range(minX, maxX + 1), (x) => {
        row += (board.has(`${x},${y}`)) ? marker : ' '
      })
      console.log(row)
    })
    console.log('')
  }
  
  // get the number of panels that have been painted at least once (counts each
  // panel once, and counts the panels even if they have been repainted black)
  return painted.size
}

(() => {
  // get input data
  const data = fs.readFileSync('../data/day11.txt')
  const inputs = parseInput(data)
  
  // Part I
  const solution1 = processInputs(inputs)
  console.log(`PART I: solution = ${solution1}`)
  
  // Part II
  const solution2 = processInputs(inputs, true, true)
  console.log('PART II: solution (see above in the shell)')
})()
